nixos/kubernetes: Module refactor

johanot opened the pull request:
### Big note: Depends on https://github.com/NixOS/nixpkgs/pull/45567 getting merged for PKI bootstrapping to work as intended, and for tests to run clean.

###### Motivation for this change

- kubernetes v1.11 should be part of NixOS 18.09. (#43882)
- security within the cluster should be the default rather than something that needs extra configuration and manual setup. (#43395)

###### This PR includes

1. major refactoring of the nixos/kubernetes module. 
2. package upgrade of kubernetes: 1.10.5 -> 1.11.2
3. manual chapter for kubernetes and release notes

The primary focus of the refactoring has been to enforce RBAC (role-based-access-control) and TLS-encryption across all cluster components. This turned out to be a cumbersome task, especially due to secret bootstrapping and exchange, which is not a trivial task. I ended up picking a solution similar to how kubeadm does cluster bootstrap. For any node to successfully join a cluster, a shared key must be manually (by an admin) copied from the master to the node, before the required node certificates can be signed and issued by the master. This is to ensure that any secrets don't end up in the store.

Secondary focus has been to make it easier to enable and disable individual cluster components independently and to make the use of `services.kubernetes.roles` more intuitive and (most importantly) optional. 

Unfortunately, upgrade from kube-dns to coredns did not happen as part of this PR. I suspect this will not make it to 18.09, unless somebody finds time to change and test. Similarly, the change from cmdline params to config files will have to wait to a later round. 

I hope that this refactoring makes the foundation for further improvements and updates to the module in the future.

fixes #43882, fixes #43395

Thanks to @srhb for much appreciated help and advise during development of this!

@kalbasit @colemickens @steveeJ Can I ask you guys for review help? 
I have a GCE cluster running 1 master and 2 nodes, for which you can get testing/playground access on request.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Commit:
nixos/kubernetes: major module refactor

- All kubernetes components have been seperated into different files
- All TLS-enabled ports have been deprecated and disabled by default
- EasyCert option added to support automatic cluster PKI-bootstrap
- RBAC has been enforced for all cluster components by default
- NixOS kubernetes test cases make use of easyCerts to setup PKI

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45670&attempt_id=d6d25e48-1336-456c-bb1a-9be8fc3fe7a8)

Attempted: kubernetes

<details><summary>Partial log (click to expand)</summary><p>

```
strip is /nix/store/y4ymnvgxygpq05h03kyzbj572zmh6zla-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/x21r9m4rdzp5pn17l8x144gnfi9hwc0s-kubernetes-1.11.2-man
checking for references to /build in /nix/store/x21r9m4rdzp5pn17l8x144gnfi9hwc0s-kubernetes-1.11.2-man...
shrinking RPATHs of ELF executables and libraries in /nix/store/m9h4y077dnrfkj0axbqq4dxhcqbbds5z-kubernetes-1.11.2-pause
shrinking /nix/store/m9h4y077dnrfkj0axbqq4dxhcqbbds5z-kubernetes-1.11.2-pause/bin/pause
strip is /nix/store/y4ymnvgxygpq05h03kyzbj572zmh6zla-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/m9h4y077dnrfkj0axbqq4dxhcqbbds5z-kubernetes-1.11.2-pause/bin
patching script interpreter paths in /nix/store/m9h4y077dnrfkj0axbqq4dxhcqbbds5z-kubernetes-1.11.2-pause
checking for references to /build in /nix/store/m9h4y077dnrfkj0axbqq4dxhcqbbds5z-kubernetes-1.11.2-pause...
/nix/store/aqd31bgwzwkpwib3ib9hfympzhcj5bx7-kubernetes-1.11.2
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45670&attempt_id=a84cfadf-0fc7-4774-9308-76538e99a63f)

Attempted: kubernetes

<details><summary>Partial log (click to expand)</summary><p>

```
strip is /nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/iddclpp6ck5g825in0dh30kavj2ddmrc-kubernetes-1.11.2-man
checking for references to /build in /nix/store/iddclpp6ck5g825in0dh30kavj2ddmrc-kubernetes-1.11.2-man...
shrinking RPATHs of ELF executables and libraries in /nix/store/7vzyfsgpw4yjkr53lp9bcyz97s0p2s2f-kubernetes-1.11.2-pause
shrinking /nix/store/7vzyfsgpw4yjkr53lp9bcyz97s0p2s2f-kubernetes-1.11.2-pause/bin/pause
strip is /nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/7vzyfsgpw4yjkr53lp9bcyz97s0p2s2f-kubernetes-1.11.2-pause/bin
patching script interpreter paths in /nix/store/7vzyfsgpw4yjkr53lp9bcyz97s0p2s2f-kubernetes-1.11.2-pause
checking for references to /build in /nix/store/7vzyfsgpw4yjkr53lp9bcyz97s0p2s2f-kubernetes-1.11.2-pause...
/nix/store/jsrwkfbcaqgy7l0njixm00xv5n7mcxb2-kubernetes-1.11.2
```
</p></details>

steveeJ commented:
@johanot thank you very much for this effort. I will take a look at this tomorrow, I'm especially curious about the TLS key generation and deployment.

johanot commented:
@steveeJ Thanks. Yeah, If I were you, I would concentrate the effort on `pki.nix`, which handles all of the TLS-stuff. Much of what happens in all the cluster component files haven't changed that much. Just me polluting the diff, because I moved a lot of code around. Sorry about that.

kristoff3r commented:
Thanks for the work!

I just tried it out using nixops, and most stuff seems to work well. The only issue so far is that I have to manually restart etcd and flannel after reboots, but I'm not sure if that is an issue with this pull request or something else.

johanot commented:
@kristoff3r Wait.. What? In the test you mean? AFAIK, I only hard-wrapped kubectl in the NixOS test case, not in the module itself? (hopefully). Thanks for testing, btw :)

kristoff3r commented:
> @kristoff3r Wait.. What? In the test you mean? AFAIK, I only hard-wrapped kubectl in the NixOS test case, not in the module itself? (hopefully). Thanks for testing, btw :)

I just figured that out as well, it was just me being confused.

grahamc commented:
cc @srhb for a review please!

srhb commented on nixos/modules/services/cluster/kubernetes/apiserver.nix:
```diff
@@ -0,0 +1,427 @@
+    };
+
+    extraSANs = mkOption {
+      description = "Extra x509 Subject Alternative Names to be added to the kubernetes apiserver tls cert.";
```
Can you add a description of what the SAN is mapped to in Kubernetes? Ie. what does it *mean* if I add another SAN "foo" here?

srhb replied:
Great, thanks!

srhb commented on nixos/modules/services/cluster/kubernetes/apiserver.nix:
```diff
@@ -0,0 +1,427 @@
+    };
+
+    featureGates = mkOption {
+      description = "List set of feature gates";
```
"List set"

srhb commented on nixos/modules/services/cluster/kubernetes/apiserver.nix:
```diff
@@ -0,0 +1,427 @@
+    serviceAccountKeyFile = mkOption {
+      description = ''
+        Kubernetes apiserver PEM-encoded x509 RSA private or public key file,
+        used to verify ServiceAccount tokens. By default tls private key file
```
ie. `tlsKeyFile`? If so, mention it.

srhb commented on nixos/modules/services/cluster/kubernetes/default.nix:
```diff
@@ -230,556 +126,22 @@ in {
     };
 
+    easyCerts = mkOption {
+      description = "Automatically setup x509 certificates and keys for the entire cluster.";
```
Maybe this should refer to the documentation for a quick rundown on how it works?

srhb commented on nixos/modules/services/cluster/kubernetes/default.nix:
```diff
@@ -230,556 +126,22 @@ in {
+
     featureGates = mkOption {
-      description = "List set of feature gates";
+      description = "List set of feature gates.";
```
Another "list set"

srhb commented on nixos/modules/services/cluster/kubernetes/pki.nix:
```diff
@@ -0,0 +1,376 @@
+    enable = mkEnableOption "Whether to enable easyCert issuer service.";
+
+    certs = mkOption {
+      description = "List of certificate specs to feed to cert generator.";
```
Can we refer the user to the specs docs here?

srhb commented on nixos/modules/services/cluster/kubernetes/pki.nix:
```diff
@@ -0,0 +1,376 @@
+
+    caSpec = mkOption {
+      description = "Certificate specification for the auto-generated CAcert.";
+      default = {
```
This is a funky default. Could we add some information that makes it clear to someone who inspects the cert that it's autogenerated?

johanot replied:
Is this better? https://github.com/NixOS/nixpkgs/pull/45670/commits/8e4070845a3ba422bc7e723db72b284f3beb8c86

srhb replied:
Yes, thank you.

srhb commented on nixos/modules/services/cluster/kubernetes/pki.nix:
```diff
@@ -0,0 +1,376 @@
+  in
+  {
+
+    services.cfssl = mkIf (top.apiserver.enable) {
```
This reminds me, do we have a warning not to try and do multimaster setup using `roles`?

srhb commented on nixos/modules/services/cluster/kubernetes/scheduler.nix:
```diff
@@ -0,0 +1,92 @@
+    };
+
+    featureGates = mkOption {
+      description = "List set of feature gates";
```
Another "list set" -- I guess this was already around before your changes

srhb commented on nixos/tests/kubernetes/base.nix:
```diff
@@ -42,34 +46,27 @@ let
               services.flannel.iface = "eth1";
-              services.kubernetes.apiserver.advertiseAddress = master.ip;
+              services.kubernetes = {
+                featureGates = ["AllAlpha"];
```
Why this gate?

johanot replied:
Removed in https://github.com/NixOS/nixpkgs/pull/45670/commits/b6ea77f764b00cedd4e7783ccb9b0a87793d1b72 . Some remnant from the days where the test depended on alpha features, I guess. It doesn't anymore.

srhb commented on nixos/doc/manual/configuration/kubernetes.xml:
```diff
@@ -0,0 +1,127 @@
+  addonManager.enable = true;
+  proxy.enable = true;
+  flannel.enable = true;
+};
```
... and configure them appropriately by hand

srhb commented on nixos/doc/manual/configuration/kubernetes.xml:
```diff
@@ -0,0 +1,127 @@
+
+ <note>
+  <para>
+   Multi-master (HA) clusters are not supported by the easyCerts module.
```
Ah, here is the warning. Thanks. :)

srhb commented:
One thing I hadn't thought about before: Would it make sense to add some notes on how to convert an easyCerts cluster to a non-easyCerts-cluster? I'm thinking specifically of how to deal with migrating, in case we change the way bootstrapping is done in the future. We can probably postpone it till then as well, but I'm sure some users would be interested in knowing how to do it.

srhb commented:
> The manual doesn't build now :(
Oh, right, I've run into this stage restriction before and I'm not sure it's actually easy to solve (I never looked into it much.) I suppose we can settle with wordy references for now ("... in the Kubernetes section of the NixOS manual...")

srhb commented:
Tangentially, I'm wondering if the documentation should really live in `nixos/modules/services/cluster/kubernetes/kubernetes.xml` and be referenced in meta.doc from its `default.nix`. @grahamc ?

johanot commented:
@srhb Squashed some commits, but now the wordy reference looks like this: https://github.com/NixOS/nixpkgs/pull/45670/commits/f9f6b9e932a9f7ed60be63eb9ace53ef08d0e92f#diff-51d78eb73f31123ee0e241af7086c3aaR135 :) Aaand the manual builds again.

srhb commented:
@GrahamcOfBorg build kubernetes

GrahamcOfBorg commented:
<!--REQUEST_ID=01e2d1c6-cc7d-4a4f-95c2-9af5fc2aed75-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45670&attempt_id=a73ea4e7-9136-4a45-a7bf-a587f446724f)

Attempted: kubernetes

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/zax3yglm2x8jnc0a4syxmpjkgc8ddkln-kubernetes-1.11.2
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=01e2d1c6-cc7d-4a4f-95c2-9af5fc2aed75-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45670&attempt_id=3740da59-39fe-41eb-838f-93e4503875e5)

Attempted: kubernetes

<details><summary>Partial log (click to expand)</summary><p>

```
strip is /nix/store/y4ymnvgxygpq05h03kyzbj572zmh6zla-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/pa11xijs2ljd7m0kipswvlhhiy68s8yy-kubernetes-1.11.2-man
checking for references to /build in /nix/store/pa11xijs2ljd7m0kipswvlhhiy68s8yy-kubernetes-1.11.2-man...
shrinking RPATHs of ELF executables and libraries in /nix/store/1yfgpql174330c04vd3vdqwsj21n1chf-kubernetes-1.11.2-pause
shrinking /nix/store/1yfgpql174330c04vd3vdqwsj21n1chf-kubernetes-1.11.2-pause/bin/pause
strip is /nix/store/y4ymnvgxygpq05h03kyzbj572zmh6zla-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/1yfgpql174330c04vd3vdqwsj21n1chf-kubernetes-1.11.2-pause/bin
patching script interpreter paths in /nix/store/1yfgpql174330c04vd3vdqwsj21n1chf-kubernetes-1.11.2-pause
checking for references to /build in /nix/store/1yfgpql174330c04vd3vdqwsj21n1chf-kubernetes-1.11.2-pause...
/nix/store/sb3jvydhf2i0irwgrl6sk76lxp1vfaxj-kubernetes-1.11.2
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=01e2d1c6-cc7d-4a4f-95c2-9af5fc2aed75-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45670&attempt_id=2054a90f-0835-4d6e-b72f-6fcc8d135936)

Attempted: kubernetes

<details><summary>Partial log (click to expand)</summary><p>

```
strip is /nix/store/df6k4mgdjxciy0f637lryp7c9ln7n1m3-cctools-binutils-darwin/bin/strip
stripping (with command strip and flags -S) in /nix/store/riwgh8ggzqsfvdzgrdhcq66zz5lfj8v2-kubernetes-1.11.2/bin
patching script interpreter paths in /nix/store/riwgh8ggzqsfvdzgrdhcq66zz5lfj8v2-kubernetes-1.11.2
gzipping man pages under /nix/store/hfjh3sbhlgnvk56agyp0jzam0mpnp3i1-kubernetes-1.11.2-man/share/man/
strip is /nix/store/df6k4mgdjxciy0f637lryp7c9ln7n1m3-cctools-binutils-darwin/bin/strip
patching script interpreter paths in /nix/store/hfjh3sbhlgnvk56agyp0jzam0mpnp3i1-kubernetes-1.11.2-man
strip is /nix/store/df6k4mgdjxciy0f637lryp7c9ln7n1m3-cctools-binutils-darwin/bin/strip
stripping (with command strip and flags -S) in /nix/store/2f6l49kjkylj98j3szbmka0sz8cvbh2c-kubernetes-1.11.2-pause/bin
patching script interpreter paths in /nix/store/2f6l49kjkylj98j3szbmka0sz8cvbh2c-kubernetes-1.11.2-pause
/nix/store/riwgh8ggzqsfvdzgrdhcq66zz5lfj8v2-kubernetes-1.11.2
```
</p></details>

colemickens commented:
One thing I noticed: `kube-dns` fails to start up on a single node cluster, for me, until I disable the firewall (`netwoking.firewall.enable = false`).

Should the module do this, or open the appropriate port(s)?

colemickens commented:
i think etcd and kubernetes are sharing the CA right now. That isn't a good idea since etcd doesn't do any client verification or authorization.

johanot commented:
@colemickens Disabling the firewall shouldn't be required for the cluster to work.. But granted, some additional documentation on firewall-configuration could be helpful. I don't think the module should open ports automatically. Speaking of which, the kubernetes flannel module opens vxlan ports. I think we should consider removing that, and adding a release note about it. @srhb Do you agree?

I'll look into adding some docs on firewall config. In the meantime, have a look at this diff for my own test cluster: https://github.com/johanot/machines-mirror/commit/6878aecb0cb84e66ef8add383d6ae0c4ab366030 .. In particular, note the `trustedInterfaces = ["flannel.1"];` <- this is necessarry for abritrary kube services to be reachable inside the cluster. For network isolation of incluster services, one should rely on kube-firewall or some external controller that enforces cluster network policies.

Regarding the etcd CA, you're right. My last line of defense was supposed to be that etcd only listen on loopback, but that's a lie. Flannel (on each node) needs etcd too - the apiserver is not the only client. I normally don't use flannel for my clusters, so I guess I totally overlooked that. In the future, I would like to replace flannel another CNI implemention like Calico or Kube-router that either doesn't use etcd at all, or use kubernetes crd's (apiserver as proxy) for storing state in etcd. ... But for now, I guess I'd have to try and generate a separate CA-cert for etcd.

srhb commented:
> Disabling the firewall shouldn't be required for the cluster to work.. But granted, some additional documentation on firewall-configuration could be helpful. I don't think the module should open ports automatically. Speaking of which, the kubernetes flannel module opens vxlan ports. I think we should consider removing that, and adding a release note about it. @srhb Do you agree?

I am sort of flip-flopping on this. On one hand it would be nice if the module just opened what it needs when it's unambiguous anyway. On the other hand I think silently opening ports is a bit risky. I think I am leaning towards opening them when things will not work without those ports.

johanot commented:
@colemickens it turns out to be complicated to have cfssl out-of-the-box deal with multiple CA-certs, so here's another way of handling the issue of access to etcd vs. kubernetes: https://github.com/johanot/nixpkgs/commit/7614c56dbddbe48921b22cce8020df079db1ca68 - with etcd rbac enabled. Let me here what you think.

GrahamcOfBorg commented:
<!--REQUEST_ID=4e7212d7-e6be-431b-acf7-ebecc80b5c16-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45670&attempt_id=1f41b674-8e98-47fb-9d2b-20ba37769b11)

Attempted: kubernetes

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/zax3yglm2x8jnc0a4syxmpjkgc8ddkln-kubernetes-1.11.2
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=4e7212d7-e6be-431b-acf7-ebecc80b5c16-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45670&attempt_id=24601fd9-fd70-480e-95d7-606fdd5ed36f)

Attempted: kubernetes

<details><summary>Partial log (click to expand)</summary><p>

```
strip is /nix/store/y4ymnvgxygpq05h03kyzbj572zmh6zla-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/pa11xijs2ljd7m0kipswvlhhiy68s8yy-kubernetes-1.11.2-man
checking for references to /build in /nix/store/pa11xijs2ljd7m0kipswvlhhiy68s8yy-kubernetes-1.11.2-man...
shrinking RPATHs of ELF executables and libraries in /nix/store/1yfgpql174330c04vd3vdqwsj21n1chf-kubernetes-1.11.2-pause
shrinking /nix/store/1yfgpql174330c04vd3vdqwsj21n1chf-kubernetes-1.11.2-pause/bin/pause
strip is /nix/store/y4ymnvgxygpq05h03kyzbj572zmh6zla-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/1yfgpql174330c04vd3vdqwsj21n1chf-kubernetes-1.11.2-pause/bin
patching script interpreter paths in /nix/store/1yfgpql174330c04vd3vdqwsj21n1chf-kubernetes-1.11.2-pause
checking for references to /build in /nix/store/1yfgpql174330c04vd3vdqwsj21n1chf-kubernetes-1.11.2-pause...
/nix/store/sb3jvydhf2i0irwgrl6sk76lxp1vfaxj-kubernetes-1.11.2
```
</p></details>

steveeJ commented:
> I normally don't use flannel for my clusters, so I guess I totally overlooked that. In the future, I would like to replace flannel another CNI implemention like Calico or Kube-router that either doesn't use etcd at all, or use kubernetes crd's (apiserver as proxy) for storing state in etcd. ... But for now, I guess I'd have to try and generate a separate CA-cert for etcd.

AFAIK flannel has gained the possibility to talk to the k8s API. PTAL at https://github.com/coreos/flannel/blob/master/Documentation/kubernetes.md#kube-flannelyaml

Commit:
nixos/kubernetes: adding manual section for kubernetes and writing release note for NixOS 19.03

fadenb commented:
A week has passed and I have used this to spin up a few hundred (single node) k8s VMs.
Except for the superficial issue that on the first boot some services fail to start (as the bootstrapping is not yet done) it worked fine 👍

MasseGuillaume commented:
@johanot Awesome work on this PR! Is there anything I can do to get this PR merged?

knedlsepp commented:
@matthewbauer Do you think this could make it for 19.03?

Commit:
nixos/kubernetes: CoreDNS privileges has to be assigned by addon manager bootstrap

- because the kube-addon-manager drops most of its privileges after
startup.

Commit:
nixos/kubernetes: use the certmgr-selfsigned variant

johanot commented:
@MasseGuillaume @knedlsepp Thanks. I've just rebased with master and fixed a minor issue that turned up in the process of getting up-to-date with the rest of nixpkgs. There's one more thing I'd like to get included in here before merging: That is, letting the etcd certs be signed by a different CA than k8s certs. As of now, you can sort of circumvent kubernetes rbac by using your k8s cert to auth against etcd - as pointed out by @colemickens. I did an experiment with implementing full etcd rbac, but that is waaay clunkier than just having separated certificate chains (or so I hope).

MasseGuillaume commented:
@johanot how do you test a service on your local machine? I tried to import it and blacklist the existing kubernetes, but the name clashes with the existing service when it loads pki.

tomberek commented:
How do you re-issue all the certificates after changing settings/IP/hostname/etc?

MasseGuillaume commented:
Ok I figured out how to test: `sudo nixos-rebuild switch -I nixpkgs=/home/gui/nixpkgs`  where nixpkgs is this repo/branch.

Howerver, the easyCert is not working for me. Here is my config:

per @tomberek comment above, localhost wont work, so I set an alternative name:

```nix
  networking = {
    extraHosts = "127.0.0.1 www.local";
  };

  kubernetes = {
     roles = ["master" "node"];
     easyCerts = true;
     masterAddress = "www.local";
    };
```

```
Feb 12 10:36:17 bob kfnkvmila18js53qszd68zby70x0sxmb-unit-script-certmgr-pre-start[13553]: 2019/02/12 10:36:17 [ERROR] cert: failed to fetch remote CA: {"code":7400,"message":"failed POST to https://www.local:8888/api/v1/cfssl/info: Post https://www.local:8888/api/v1/cfssl/info: dial tcp 127.0.0.1:8888: connect: connection refused"}
Feb 12 10:36:17 bob kfnkvmila18js53qszd68zby70x0sxmb-unit-script-certmgr-pre-start[13553]: Failed: {"code":7400,"message":"failed POST to https://www.local:8888/api/v1/cfssl/info: Post https://www.local:8888/api/v1/cfssl/info: dial tcp 127.0.0.1:8888: connect: connection refused"}
Feb 12 10:36:17 bob systemd[1]: certmgr.service: Control process exited, code=exited status=1
Feb 12 10:36:17 bob systemd[1]: certmgr.service: Failed with result 'exit-code'.
Feb 12 10:36:17 bob systemd[1]: Failed to start certmgr.
```

I can connect  to cfssl server without any problem via curl:

```
 curl -k -XPOST https://www.local:8888/api/v1/cfssl/info
{"success":false,"result":null,"errors":[{"code":400,"message":"unexpected end of JSON input"}],"messages":[]}
```

Commit:
nixos/kubernetes: let flannel use kubernetes as storage backend

+ isolate etcd on the master node by letting it listen only on loopback
+ enabling kubelet on master and taint master with NoSchedule

The reason for the latter is that flannel requires all nodes to be "registered"
in the cluster in order to setup the cluster network. This means that the
kubelet is needed even at nodes on which we don't plan to schedule anything.

tomberek commented:
@MasseGuillaume Yeah, i did the same thing with hosts. I ran into the same issue with certmgr when iterating on settings. It's possible some TLS connections are being refused because the issued certs are with older IPs, domain names, etc - hence, my question about re-issuing. I'm doing all this with NixOps, so i end up destroy/creating a lot.

johanot commented:
@MasseGuillaume @tomberek certmgr doesn't actually verify that the certs already issued match the specs provided, see: https://github.com/cloudflare/certmgr/issues/48 .. Except the expiration date, AFAIK nothing of the actual content of the certificate is monitored, which, is just one of the reason why an improved cfssl client is high on my wishlist. You can however force a re-issue by `rm`-ing the issued certs. certmgr will detect which certs are missing and execute a reissue.

calbrecht commented:
On the way to include the metrics-server addon (needs to extend kubernetes api).
Pod of metrics-server keeps crashing because of missing requestheader-client-ca-file.

Documentation suggests a third ca cert (for kubernetes extension api to work).
https://kubernetes.io/docs/tasks/access-kubernetes-api/configure-aggregation-layer/

Now i'm going to refactor the nixos kubernetes module to setup the certificates and keys according to the docs https://kubernetes.io/docs/setup/certificates/ to be able to make more sense of warnings and errors in the logs, as well as to easier find and follow relevant kubernetes documentation.

Commit:
nixos/kubernetes: use system.path to handle dependency on flannel subnet.env

The current postStart step on flannel causes flannel.service to
sometimes hang, even when it's commanded to stop.

Commit:
nixos/kubernetes: (test) Fix race-condition in test cases. docker load might fail due to dockerd restarting

johanot commented:
Ok.. Turns out that having multiple CA's is too big of a refactor for this PR. Here's how I tried to fix the etcd security issue for now:

- Let flannel use kubernetes as configuration backend. Meaning that it communicates with kube-apiserver instead of directly with etcd.
- Because of the flannel change, we can now isolate etcd on the master node letting it listen only on loopback and closing ports 2379 and 2380 in the firewall.

Have a look at 54c09c009c96d7f1795fe9d65d37db08e3730ef7 in particular. This also implies merging of https://github.com/NixOS/nixpkgs/pull/55649 by the way.

Let me know what you think @fpletz @tomberek @MasseGuillaume @srhb @colemickens 

We'll have to revisit this later, especially when we want "easyCerts"-support for multi-master clusters. But since easyCerts only supports a single-master setup for other reasons, it's ok to isolate the single etcd instance, IMO.

johanot commented:
For the record: Here's how I test.. When developing I mostly run: `nix-build ./nixos/release.nix -A tests.kubernetes.rbac.singlenode -A tests.kubernetes.rbac.multinode -A tests.kubernetes.dns.singlenode -A tests.kubernetes.dns.multinode`. For more exhaustive integration testing I have a small GCE cluster. Here's the config: https://github.com/johanot/machines-mirror/tree/master/configurations/nixos

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/trying-to-run-kubernetes-in-a-single-node-mode/2159/3

Ma27 commented:
(removed my review request here, I obviously got notified accidentally as I didn't do anything k8s related in nixpkgs these days)

calbrecht commented:
@johanot, finally all kubernetes tests are passing for https://github.com/NixOS/nixpkgs/compare/master...mayflower:k8s-refactor and certificates are generated and signed by the multirootca service that is running on the kubernetes master. 

Could you please try it against your cluster and report any failures?
I will sort the commits until this evening.

johanot commented:
@calbrecht Thank you for your additions and improvements. I haven't tested your branch yet, but I'd had a brief overlook of the changes. In general, I feel that in order to not overburden the reviewers, we might need to split the changes over multiple PRs. Besides the metrics-server and the multirootca, I see that your PR includes a move from nix expr to json-files for addons, along with the addition the `proxy-client` flags, and general rbac changes. I think that both of our PRs are heavily overloaded and that makes them very difficult to review.

Would it be possible to add the metrics server in a separate PR instead of rebasing and refacoring this one?

srhb commented:
Yes, it's already quite a mouthful, so improvements that are tangential to the main purpose of this PR would benefit from coming in later, I feel. Let's prioritize getting to a place where active NixOS k8s users can agree on a state as quickly as possible (but no quicker!) in which this feels safe and right to merge.

fpletz reviewed (approved):
@calbrecht and me just reviewed the new changes for flannel and etcd and inspected the cluster state in the NixOS tests. We think this should be merged ASAP because it makes the k8s NixOS module usable for quick test deploys of a k8s cluster. We will open new PRs for @calbrecht changes after this PR is merged.

As we're approaching the 19.03 feature freeze, I would like to merge this in the current state in 24 hours if nobody disagrees. From the feedback here and Discourse it looks like quite a few people are already running or have tested this PR.

johanot commented:
@fpletz Great! Thanks. Wan't me to do some squashing/rebasing, or will you handle that?

johanot commented:
@fpletz Rebased with master now.

calbrecht commented:
Yes of course you are both right. I should have been mentioning, which parts of the branch i see as relevant for this PR.

The order of systemd service execution i see as highly relevant, because currently log files are clobbered with failures because of dependent services did not start and e.g. execution of tests is taking long because of the restarts (That accounts for a simple deploy or reboot as well). Adding order is a cheap win, imho.

The change from cfssl to multirootca i see as relevant as well, i just can't explain it well because of missing knowledge and lack of terminology. Naively speaking, it fills the gap to deliver the security purpose of mutual TLS secured communication when relying on easyCerts.

The other changes are indeed not related to this PR and stem from me getting into running helm chart deployments on the nixos kubernetes module cluster.

offlinehacker commented:
I'm using this now in tests for `kubenix` 2.0 and works great, thanks :)

offlinehacker commented:
On every so often the tests breaks cycling with following error. I'm running tests with cpu count set to `$(nproc)`, so it might be a race condition between cni and docker, any ideas?

```
kubelet[2203]: E0220 17:55:50.116782    2203 kuberuntime_sandbox.go:68] CreatePodSandbox for pod "nginx-8956df8db-x9zb4_default(7ddd7fb8-3538-11e9-b977-525400123456)" failed: rpc error: code = Unknown desc = [failed to set up sandbox container "d2a87a386e3ba5a616f1917f53d5e8cdfd810e0806acd2445b9fbaa955181106" network for pod "nginx-8956df8db-x9zb4": NetworkPlugin cni failed to set up pod "nginx-8956df8db-x9zb4_default" network: failed to set bridge addr: "docker0" already has an IP address different from 10.1.0.1/24, failed to clean up sandbox container "d2a87a386e3ba5a616f1917f53d5e8cdfd810e0806acd2445b9fbaa955181106" network for pod "nginx-8956df8db-x9zb4": NetworkPlugin cni failed to teardown pod "nginx-8956df8db-x9zb4_default" network: running [/nix/store/0a9hwaxd5vln0h9j7rkcvh8kmzc41dfm-iptables-1.8.2/bin/iptables -t nat -D POSTROUTING -s 10.1.0.98/24 -j CNI-2f2a5e44feafab3d2782369f -m comment --comment name: "mynet" id: "d2a87a386e3ba5a616f1917f53d5e8cdfd810e0806acd2445b9fbaa955181106" --wait]: exit status 2: iptables v1.8.2 (legacy): Couldn't load target `CNI-2f2a5e44feafab3d2782369f':No such file or directory
kube# [  161.400595] kubelet[2203]: Try `iptables -h' or 'iptables --help' for more information.
```

offlinehacker commented:
On other local machine, where i'm running nested virtualization and everything is running really slow, i'm getting following error when running test. Yes i named VM running test `kube`

```
kube: running command: kubectl get node kube.my.zyx | grep -w Ready
kube# Error from server (NotFound): nodes "kube.my.zyx" not found
kube: exit status 1
(0.35 seconds)
kube# [  203.696806] flannel[3510]: I0220 18:09:21.325902    3510 kube.go:133] Node controller sync successful
kube# [  203.712007] flannel[3510]: I0220 18:09:21.326098    3510 main.go:244] Created subnet manager: Kubernetes Subnet Manager - kube.my.zyx
kube# [  203.736696] flannel[3510]: I0220 18:09:21.326228    3510 main.go:247] Installing signal handlers
kube# [  203.751432] flannel[3510]: I0220 18:09:21.328038    3510 main.go:386] Found network config - Backend type: vxlan
kube# [  203.763631] flannel[3510]: I0220 18:09:21.328205    3510 vxlan.go:120] VXLAN config: VNI=1 Port=0 GBP=false DirectRouting=false
kube# [  203.771951] flannel[3510]: E0220 18:09:21.328495    3510 main.go:289] Error registering network: failed to acquire lease: node "kube.my.zyx" not found
kube# [  203.789721] flannel[3510]: I0220 18:09:21.335167    3510 main.go:366] Stopping shutdownHandler...
kube# [  203.802968] systemd[1]: flannel.service: Main process exited, code=exited, status=1/FAILURE
```

fpletz commented:
We've just observed similar errors while giving the Nixos tests another spin on some other machines. It depends on the number of cores your build machine has and how many tests you run in parallel whether some of the tests succeed. We expect that @calbrecht's work on the service dependencies ordering will make the tests stable.

We would nevertheless like to merge this now and fix the remaining issues before more conflicts with master arise.

johanot commented:
@calbrecht @fpletz Can you make a separate PR with only the dependency changes? I feel that those are kind of a must-have have for 19.03.

fpletz commented:
@johanot That was the plan. Definitely before 19.03. :)

fpletz committed:
nixos/kubernetes: bootstrap docker without networking

Before flannel is ready there is a brief time where docker will be
running with a default docker0 bridge. If kubernetes happens to spawn
containers before flannel is ready, docker can't be restarted when
flannel is ready because some containers are still running on the
docker0 bridge with potentially different network addresses.

Environment variables in `EnvironmentFile` override those defined via
`Environment` in the systemd service config.

Co-authored-by: Christian Albrecht <christian.albrecht@mayflower.de>

fpletz commented:
Added a fix for the docker/flannel bootstrap issue.

fpletz commented:
Rebased on top of current master to make @GrahamcOfBorg happy. :smile:

offlinehacker commented:
After i added `Restart="on-failure";` to kubelet it works fine, @fpletz can you add auto restart on failure to kubelet, as this is what you usually want? We also have same on docker.

offlinehacker commented:
There's also an issue with `kubelet-boostrap` failing during tests due docker restarting in between load, do you have any idea how to fix this?

offlinehacker commented:
Ok, i made these changes and this seems to fix all race conditions: https://github.com/GateHubNet/nixpkgs/commit/86e544bd3656a396050325e8853c3a40d1611427

offlinehacker commented:
Also single node tests do not need flannel, but this can be improved later

Commit:
nixos/kubernetes: fix flannel and kubelet startup

fpletz commented:
@offlinehacker Awesome! Seems to work on a few of our machines. We've run the tests a couple of times.

We've found some other issues while trying to address your comments that we will incoroprate in the following PRs.

Cherry-picked your commit. Time to merge! :rocket:

Thanks to all contributors here! :beers:

MasseGuillaume commented:
Hi @johanot,

I'm trying to use this PR to set up my developer environment.

I saw in the documentation that it's possible to login as the cluster admin with the key that was generated.
https://github.com/NixOS/nixpkgs/blob/2d20e8c5f2e029606e65d2e480b5d1d961c69f50/nixos/doc/manual/configuration/kubernetes.xml#L114-L125

Would it be possible to add some documentation on how to create cert/key for a user? This is what I got so far:

```nix
{
environment.systemPackages = 
    let
      kub = config.services.kubernetes;
      dev = kub.lib.mkCert {
        name = "gui";
        CN = "gui";
        fields = {
          O = "system:masters";
        };
        privateKeyOwner = "gui";
      };
      kubeConfig = kub.lib.mkKubeConfig "gui" {
        server = kub.apiserverAddress;
        certFile = dev.cert;
        keyFile = dev.key;
      };
      kubectl2 = with pkgs; runCommand "wrap-kubectl" { buildInputs = [ makeWrapper ]; } ''
        mkdir -p $out/bin
        echo "---"
        echo ${dev.cert}
        echo ${dev.key}
        echo "---"
        echo "makeWrapper ${pkgs.kubernetes}/bin/kubectl $out/bin/kubectl --set KUBECONFIG "${kubeConfig}""
        makeWrapper ${pkgs.kubernetes}/bin/kubectl $out/bin/kubectl2 --set KUBECONFIG "${kubeConfig}"
      '';
    in [
    kubectl2
  ]
}
```

However dev.cert and dev.key points to non existing files.

Is this the correct approach ?

arianvp commented on nixos/modules/services/cluster/kubernetes/pki.nix:
```diff
@@ -0,0 +1,388 @@
+        echo "Node joined succesfully"
+      '')];
+
+      # isolate etcd on loopback at the master node
```
It would be nice if easyCerts could be used in conjunction with an external etcd cluster. Now this silently disables that possibility without telling the user.   We should probably document this behaviour with **big fat letters** somewhere

johanot replied:
@arianvp couldn't you do something like:
```
  services.etcd.enable = pkgs.lib.mkForce false;
  services.kubernetes.apiserver.etcd = {
    servers = [ "some.other.server.org" ];
    keyFile = "/some/key.pem";
    certFile = "/some/cert.pem";
    caFile = "/some/ca.pem";
  };
```
Or am I missing some details here? (The `mkForce` is due to a legacy thing with the flannel module, not easyCerts).

I don't agree with the *silently disables* phrasing. easyCerts is a feature you can opt-in to (or choose not to), which automates almost every part of the setup of a small basic Kubernetes cluster. It was never meant to satisfy every need, nor was it designed to make cluster customizations easy. It doesn't make customization impossible though - just less automatic. :-) I don't know if the current documentation of easyCerts needs a disclaimer of some sort, but feel free to PR both module and docs.

arianvp replied:
Thanks! Makes sense.

johanot commented:
@MasseGuillaume As far as I can see, you miss the final step of adding your certificate spec to the pki module (ultimately certmgr), which will then issue your certificates. Something along the lines of: `config.services.kubernetes.pki.certs = { inherit dev; }`. Yeah we should perhaps extend the docs to cover how to add extra custom certs.

MasseGuillaume commented:
@johanot Thanks I will try this and send a PR to update the documentation. Once again, thanks for your hard work!

MasseGuillaume commented:
@johanot I got the following to work:

```nix
{ config, pkgs, ... }:

let
  kub = config.services.kubernetes;
  devCert = kub.lib.mkCert {
    name = "gui";
    CN = "kubernetes-cluster-ca";
    fields = {
      O = "system:masters";
    };
    privateKeyOwner = "gui";
  };
  kubeConfig = kub.lib.mkKubeConfig "gui" {
    server = kub.apiserverAddress;
    certFile = devCert.cert;
    keyFile = devCert.key;
  };
in
{
  services = {
    kubernetes = {
      roles = ["master" "node"];
      easyCerts = true;
      masterAddress = "localhost";
      pki.certs = { dev = devCert; };
    };
  }

  programs.zsh = {
    interactiveShellInit = ''
      export KUBECONFIG="${kubeConfig}:$HOME/.kube/config";
    '';
  }
}
```

luaduck commented:
Would it be possible to implement a better way of configuring kube-addon-manager without having to create kubeconfig ourselves if not using easyCert?

I'm talking about [pki.nix#238](https://github.com/NixOS/nixpkgs/pull/45670/files#diff-8d468cb36655335308a8eb4c3999b59aR238) - from what I can see, unless easyCert is in control of certificates, I have no choice but to call `services.kubernetes.lib.mkKubeConfig` myself with my own certificate paths.

Maybe bring kubeconfig generation into the configuration inside `services.kubernetes.addonManager.*`, ala the other modules? (i.e `services.kubernetes.addonManager.keyFile`, etc.)

I could be missing something obvious here, so I'm fine with being proved blind!
